### DISCUSSION TITLE
Is ARM motion planner path always to generate in front of the cloths?

### DISCUSSION THREAD

**Discussion by nero8919** [Victorlouisdg/cloth-competition#18] [Questions] [answered]
Hello I'm JungHyun Choi from Uos-robotics
Thank you for your efforts in running the competition.

I read the code 'planng/grasp_planning.py' and it generate trajectory with airo_drake.
It seems to calculate no-collision shortest path with arm and cloths. 

I think that if a grasp pose is set to the awkward pose , which z-axis of gripper is heading for right side, nearby the left arm, the robot can arrive at that pose in different paths(front or back of cloth). 

Even if the robot is holding the same point, going behind the cloths and grabbing it may lead to different results in the stretching process if it goes behind the cloths and grabs it in front of the cloths.
 
Could you please provide some insights or clarifications on how the algorithm determines the robot's approach direction in such scenarios?

![temppose](https://github.com/Victorlouisdg/cloth-competition/assets/74824730/8139e35c-ee17-49fb-b704-96d727e156d2)

**Comment by Victorlouisdg** [comment, marked as answer]
Hi JungHyun Choi

Thank you for your excellent question. I must apologize for not being able to clean up and document the motion planning code more, but will try my best to explain it here.

You are correct that there are multiple ways to approach the cloth to grasp it and to stretch it afterward. The solution chosen by motion planning can indeed influence the final result. We tried to make as many grasp poses as possible around the cloth feasible, but as you noted this can be challenging for certain scenarios (e.g. the one you describe where the right gripper points right). We also tried to ensure that the paths taken lead to optimal unfolding for the chosen grasp pose, but there is certainly still room for improvement.

The reason that there are multiple solutions comes down to two reasons:

1.  For a single goal TCP pose within the workspace of a UR5e robot, there can be up to 8 inverse kinematics joint solutions.
2. Toward these 8 joint solutions, often many paths can be taken.

For finding and choosing paths, we use airo-planner which internally uses RRTConnect from OMPL. We search for collision-free paths and also try to consider execution time, twisting of the cloth, and strain on the cloth. RRTConnect is a sampling-based motion planner, so there is randomness in which solution path it will find.

I'd love to explain how the motion planning for the grasping and stretching works in more detail, but I'm a bit short on time right now. However, I'll definitely provide a more detailed explanation next week! In the meantime, feel free to ask any follow-up questions you have. I'll do my best to answer them alongside the longer explanation next week.
